include/annotated.hpp

```diff
@@ -92,6 +93,74 @@ using DefaultSupportedTypes = SupportedTypes<>;
 static_assert(fair::meta::is_instantiation_of<DefaultSupportedTypes, SupportedTypes>);
 static_assert(fair::meta::is_instantiation_of<SupportedTypes<float, double>, SupportedTypes>);
 
+/**
+ * @brief Represents limits and optional validation for an Annotated<..> type.
+ *
+ * The `Limits` structure defines lower and upper bounds for a value of type `T`.
+ * Additionally, it allows for an optional custom validation function to be provided.
+ * This function should take a value of type `T` and return a `bool`, indicating
+ * whether the value passes the custom validation or not.
+ *
+ * Example:
+ * ```
+ * Annotated<float, "example float", Visible, Limits<0.f, 1024.f>>             exampleVar1;
+ * // or:
+ * constexpr auto isPowerOfTwo = [](const int &val) { return val > 0 && (val & (val - 1)) == 0; };
+ * Annotated<float, "example float", Visible, Limits<0.f, 1024.f, isPowerOfTwo>> exampleVar2;
+ * // or:
+ * Annotated<float, "example float", Visible, Limits<0.f, 1024.f, [](const int &val) { return val > 0 && (val & (val - 1)) == 0; }>> exampleVar2;
+ * ```
+ */
+template<auto LowerLimit, decltype(LowerLimit) UpperLimit, auto Validator = nullptr>
+    requires(requires(decltype(Validator) f, decltype(LowerLimit) v) {
+        { f(v) } -> std::same_as<bool>;
+    } || Validator == nullptr)
+struct Limits {
+    using ValueType = decltype(LowerLimit);
+    static constexpr ValueType MinRange = LowerLimit;
+    static constexpr ValueType MaxRange = UpperLimit;
+    static constexpr decltype(Validator) ValidatorFunc = Validator;
+
+    static constexpr bool
+    validate(const ValueType &value) noexcept {
+        if (LowerLimit == UpperLimit) {
+            if constexpr (Validator != nullptr) {
+                try {
+                    return Validator(value);
+                } catch (...) {
+                    return false;
+                }
+            } else {
+                return true; // if no validator and limits are same, return true by default
+            }
+        } else {
+            if constexpr (Validator != nullptr) {
+                try {
+                    if (value < LowerLimit || value > UpperLimit || !Validator(value)) return false;
+                } catch (...) {
+                    return false;
+                }
+            } else {
+                if (value < LowerLimit || value > UpperLimit) return false;
+            }
+            return true;
+        }
+    }
+};
```

Comment:
Is there a reason for nesting the otherwise not related branches? The following should be identical and save some nesting and repetition.

``` c++
if (LowerLimit == UpperLimit) {
    if (value < LowerLimit || value > UpperLimit) return false;
}
if constexpr (Validator != nullptr) {
    try {
        if (!Validator(value)) return false;
    } catch (...) {
        return false;
    }
}
return true;
```

Reply:
Note the `constexpr-if' cascade ... your code above is runtime evaluated.

Reply:
...pushed the wrong version ... sorry. fixed.